src/Zend/Request.php

```diff
@@ -43,7 +44,8 @@ public function __construct(
         $this->setAllowCustomMethods(true);
 
         $this->setMethod($method);
-        $this->setRequestUri((string) $uri);
+        // Remove the "http(s)://hostname" part from the URI
+        $this->setRequestUri(preg_replace('#^[^/:]+://[^/]+#', '', (string) $uri));
```

Comment:
Probably a better approach would be:

```php
$normalizedUri = $uri->getPath();
$normalizedUri .= $uri->getQuery() ? '?' . $uri->getQuery() : '';
$normalizedUri .= $uri->getFragment() ? '#' . $uri->getFragment() : '';
$this->setRequestUri($normalizedUri);
```

This will better ensure that if there is a missing scheme or the authority contains any unexpected characters, you will still have a URI that works correctly with zend-http.

Reply:
We used the regular expression because it's already used in `Zend\Http\PhpEnvironment\Request` [see here](https://github.com/zendframework/zend-http/blob/master/src/PhpEnvironment/Request.php#L464).
Moreover, are you sure we need to include the fragment part?

Reply:
In the regex you provide above, the fragment will be included. If you want to omit it, my approach would make that simpler to accomplish.

Reply:
@weierophinney Your solution can be applied only if `$uri` is instanceof `Psr\Http\Message\UriInterface` but it can be also a string. I think the regexp is a more general solution here. What do you think?

Reply:
Ah, I was unaware that it could be a string; I assumed it was a `UriInterface` instance.

So... carry on!